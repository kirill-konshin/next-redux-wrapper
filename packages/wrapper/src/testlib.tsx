import {AppContext} from 'next/app';
import {applyMiddleware, createStore, AnyAction} from 'redux';
import promiseMiddleware from 'redux-promise-middleware';
import React, {Component} from 'react';
import renderer from 'react-test-renderer';

export const reducer = (state = {reduxStatus: 'init'}, action: AnyAction) => {
    switch (action.type) {
        case 'FOO': // sync
        case 'FOO_FULFILLED': // async
            return {reduxStatus: action.payload};
        default:
            return state;
    }
};

export const makeStore = (initialState: any) => createStore(reducer, initialState, applyMiddleware(promiseMiddleware));

class SyncPageBase extends Component<any> {
    public render() {
        const {store, ...props} = this.props;
        return (
            <div>
                {JSON.stringify(props)}
                {JSON.stringify(store.getState())}
            </div>
        );
    }
}

export class SyncPage extends SyncPageBase {
    public static getInitialProps({ctx}: AppContext) {
        ctx.store.dispatch({type: 'FOO', payload: 'foo'});
        return {custom: 'custom'};
    }
}

const someAsyncAction = {
    type: 'FOO',
    payload: new Promise(res => res('foo')),
};

export class AsyncPage extends SyncPageBase {
    public static async getInitialProps({ctx}: AppContext) {
        await ctx.store.dispatch(someAsyncAction);
        return {custom: 'custom'};
    }
}

export class NoStorePage extends SyncPageBase {
    public static async getInitialProps() {
        return {custom: 'custom'};
    }
}

/**
 * Creates a stub AppContext declared as any
 */
export function createAppContext(): any {
    return {
        ctx: {
            pathname: '/',
            req: {},
            res: {},
            query: {},
        },
    };
}

export async function verifyComponent(WrappedPage: any) {
    // this is called by Next.js
    const props = await WrappedPage.getInitialProps(createAppContext());

    expect(props.initialProps.custom).toBe('custom');
    expect(props.initialState.reduxStatus).toBe('foo');

    // this is called by Next.js
    const component = renderer.create(<WrappedPage {...props} />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
}
