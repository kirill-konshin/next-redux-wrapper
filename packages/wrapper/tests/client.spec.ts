/**
 * @jest-environment jsdom
 **/

import {verifyComponent, makeStore, NoStorePage, SyncPage, wrapper} from './testlib';
import {act} from 'react-dom/test-utils';

describe('client integration', () => {
    test('store taken from window', async () => {
        const WrappedPage = wrapper.withRedux(NoStorePage);
        const store = makeStore();
        (window as any)['testStoreKey'] = store;
        act(() => {
            store.dispatch({type: 'FOO', payload: 'foo'});
        });
        await verifyComponent(WrappedPage);
    });
    test('store not taken from window', async () => {
        const WrappedPage = wrapper.withRedux(SyncPage);
        await verifyComponent(WrappedPage);
    });
});
