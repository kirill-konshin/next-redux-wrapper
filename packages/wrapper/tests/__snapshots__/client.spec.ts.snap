// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client integration store not taken from window 1`] = `
<div>
  {"custom":"custom"}
  {"reduxStatus":"foo"}
</div>
`;

exports[`client integration store taken from window 1`] = `
<div>
  {"custom":"custom"}
  {"reduxStatus":"foo"}
</div>
`;
